src/block_service/block_service_manager.cpp

```diff
@@ -33,54 +33,52 @@ block_service_manager::block_service_manager()
 {
 }
 
-block_filesystem *block_service_manager::get_block_filesystem(const std::string &provider)
+block_service_manager::~block_service_manager()
 {
-    {
-        zauto_read_lock l(_fs_lock);
-        auto iter = _fs_map.find(provider);
-        if (iter != _fs_map.end())
-            return iter->second.get();
-    }
+    ddebug("close block service manager.");
+    zauto_write_lock l(_fs_lock);
+    _fs_map.clear();
+}
 
+block_filesystem *block_service_manager::get_or_create_block_filesystem(const std::string &provider)
+{
     {
         zauto_write_lock l(_fs_lock);
         auto iter = _fs_map.find(provider);
-        if (iter != _fs_map.end())
+        if (iter != _fs_map.end()) {
             return iter->second.get();
+        }
+    }
 
-        const char *provider_type = dsn_config_get_value_string(
-            (std::string("block_service.") + provider).c_str(), "type", "", "block service type");
+    const char *provider_type = dsn_config_get_value_string(
+        (std::string("block_service.") + provider).c_str(), "type", "", "block service type");
 
-        block_filesystem *fs =
-            utils::factory_store<block_filesystem>::create(provider_type, PROVIDER_TYPE_MAIN);
-        if (fs == nullptr) {
-            derror("acquire block filesystem failed, provider = %s, provider_type = %s",
-                   provider.c_str(),
-                   provider_type);
-            return nullptr;
-        }
+    block_filesystem *fs =
```

Comment:
I think it better to use `std::unique_ptr<block_filesystem>` here

Reply:
The `block_filesystem` will be inserted into `_fs_map` and returned, so maybe using `unique_ptr` is not appropriate.